tower-discover/src/lib.rs

```diff
@@ -82,6 +84,58 @@ where
     }
 }
 
+/// Dynamic service discovery based on a stream of service changes.
+pub struct Services<S, K, Svc> {
+    inner: S,
+    _marker_k: PhantomData<K>,
+    _marker_v: PhantomData<Svc>,
+    finished: bool,
```

Comment:
The `finished` logic is what `Fuse` does: https://docs.rs/futures/0.1.25/futures/future/struct.Fuse.html